packages/grid/x-data-grid-premium/src/hooks/features/clipboard/useGridClipboardImport.ts

```diff
@@ -0,0 +1,435 @@
+        focusedCell.focus({ preventScroll: true });
+      }
+      el.remove();
+      resolve(text || '');
+    };
+
+    el.addEventListener('paste', handlePasteEvent);
+    rootEl.appendChild(el);
+    el.focus({ preventScroll: true });
+  });
+}
+
+const stringToBoolean = (value: string) => {
+  switch (value.toLowerCase().trim()) {
+    case 'true':
+    case 'yes':
+    case '1':
+      return true;
+
+    case 'false':
+    case 'no':
+    case '0':
+    case 'null':
+    case 'undefined':
+      return false;
+
+    default:
+      return undefined;
+  }
+};
+
+const parseCellStringValue = (value: string, colDef: GridColDef) => {
+  switch (colDef.type) {
+    case 'number': {
+      return Number(value);
+    }
+    case 'boolean': {
+      return stringToBoolean(value);
+    }
+    case 'singleSelect': {
+      const colDefValueOptions = (colDef as GridSingleSelectColDef).valueOptions;
+      const valueOptions =
+        typeof colDefValueOptions === 'function'
+          ? colDefValueOptions({ field: colDef.field })
+          : colDefValueOptions || [];
+      const getOptionValue = (colDef as GridSingleSelectColDef).getOptionValue!;
+      const valueOption = valueOptions.find((option) => {
+        if (getOptionValue(option) === value) {
+          return true;
+        }
+        return false;
+      });
+      if (valueOption) {
+        return value;
+      }
+      return value;
+    }
+    case 'date':
+    case 'dateTime': {
+      const date = new Date(value);
+      return date;
+    }
+    default:
+      return value;
+  }
+};
+
+// Keeps track of updated rows during clipboard paste
+class CellValueUpdater {
+  rowsToUpdate: {
+    [rowId: GridRowId]: GridValidRowModel;
+  } = {};
+
+  updateRow: (row: GridRowModel) => void;
+
+  options: {
+    apiRef: React.MutableRefObject<GridPrivateApiPremium>;
+    processRowUpdate: DataGridPremiumProcessedProps['processRowUpdate'];
+    onProcessRowUpdateError: DataGridPremiumProcessedProps['onProcessRowUpdateError'];
+    getRowId: DataGridPremiumProcessedProps['getRowId'];
+  };
+
+  constructor(options: CellValueUpdater['options']) {
+    this.options = options;
+    this.updateRow = batchRowUpdates(options.apiRef.current.updateRows, 50);
+  }
+
+  updateCell({
+    rowId,
+    field,
+    pastedCellValue,
+  }: {
+    rowId: GridRowId;
+    field: GridColDef['field'];
+    pastedCellValue: string;
+  }) {
+    if (pastedCellValue === undefined) {
+      return;
+    }
+
+    const { apiRef, getRowId } = this.options;
+    const colDef = apiRef.current.getColumn(field);
+    if (!colDef || !colDef.editable) {
+      return;
+    }
+    const row = this.rowsToUpdate[rowId] || { ...apiRef.current.getRow(rowId) };
+    if (!row) {
+      return;
+    }
+
+    let parsedValue = parseCellStringValue(pastedCellValue, colDef);
+
+    if (colDef.valueParser) {
+      parsedValue = colDef.valueParser(parsedValue, apiRef.current.getCellParams(rowId, field));
+    }
```

Comment:
Would it make more sense to add `pasteValueParser` to handle pasted values specifically instead of the `reason`?
Then, for each cell, we'll call:
- `pastedValueParser` (if available) - to parse the pasted string to whatever you want
- `valueParser` (if available) - to transform the parsed value

A potential use case for this would be: using `valueParser` to reset the time when editing `dateTime` column:
```tsx
valueParser: (value) => {
  const newDate = new Date(value);
  newDate.setHours(12);
  newDate.setMinutes(0);
  return newDate;
}
```

In this case, the default `pasteValueParser` will convert pasted string to a Date object, but you can override it with a custom one. The `valueParser` stays unchanged in this case.

What do you think?

Reply:
Done!